File: arrays_and_hashing/longest_consecutive.ts
//https://leetcode.com/problems/longest-consecutive-sequence/
// TODO: Ask why is this 0(n) doenst accessing set value add more?
// TODO: Ask why while cycle doesnt add to complexity
// TODO: Solve again

function longestConsecutive(nums: number[]): number {
  if (!nums.length) return 0
  const set = new Set(nums);
  let longest = 0;

  nums.forEach(n => {
    if (set.has(n - 1)) return
    let length = 0

    while (set.has(n + length)) {
      length++
    }

    if (length > longest) longest = length

  })
  return longest
};



const test = [100, 4, 200, 1, 3, 2];
console.log(longestConsecutive(test))

File: arrays_and_hashing/go/longest-consecutive.go
// https://leetcode.com/problems/longest-consecutive-sequence/
// TODO: solve again

package main

import (
	"fmt"
)

func main() {
	test := []int{100, 4, 200, 1, 3, 2}
	fmt.Println(longestConsecutive(test))
}

func longestConsecutive(nums []int) int {
	m := make(map[int]bool)

	for _, n := range nums {
		m[n] = true
	}

	longest := 0

	for _, n := range nums {
		isBeginning := !m[n-1]

		if !isBeginning {
			continue
		}

		inc := 1
		last := m[n+inc]

		for last {
			inc++
			last = m[n+inc]
		}

		if inc > longest {
			longest = inc
		}

	}

	return longest
}
